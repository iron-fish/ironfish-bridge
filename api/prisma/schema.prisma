generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-3.0.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model BridgeRequest {
  id                      Int                   @id @default(autoincrement())
  asset                   String                @db.VarChar
  source_address          String                @db.VarChar
  destination_address     String                @db.VarChar
  source_transaction      String?               @unique @db.VarChar
  destination_transaction String?               @db.VarChar
  source_chain            Chain
  destination_chain       Chain
  status                  BridgeRequestStatus
  amount                  String                @db.VarChar
  failure_reason          FailureReason?
  completed_at            DateTime?             @db.Timestamp(6)
  created_at              DateTime              @default(now()) @db.Timestamp(6)
  started_at              DateTime?             @db.Timestamp(6)
  updated_at              DateTime              @default(now()) @updatedAt @db.Timestamp(6)
  wiron_burn_transaction  String?               @db.VarChar
  failures                FailedBridgeRequest[]

  @@index([source_address], map: "index_bridge_request_on_source_address")
  @@index([destination_address], map: "index_bridge_request_on_destination_address")
  @@index([wiron_burn_transaction], map: "index_bridge_request_wiron_burn_transaction")
  @@index([completed_at], map: "index_bridge_requests_on_completed_at")
  @@index([started_at, completed_at], map: "index_bridge_requests_on_completed_at_and_started_at")
  
  @@map("bridge_requests")
}

model FailedBridgeRequest {
  id                Int            @id @default(autoincrement())
  bridge_request_id Int?
  failure_reason    FailureReason
  error             String?        @db.VarChar
  bridge_request    BridgeRequest? @relation(fields: [bridge_request_id], references: [id])

  @@index([bridge_request_id], map: "index_failed_bridge_request_on_bridge_request_id")
  @@map("failed_bridge_requests")
}

model IronFishTestnetHead {
  hash String @unique @db.VarChar

  @@map("iron_fish_testnet_head")
}

model WIronSepoliaHead {
  id     Int    @id
  hash   String @db.VarChar
  height Int

  @@map("wiron_sepolia_head")
}

enum Chain {
  IRONFISH
  ETHEREUM

  @@map("chain_name")
}

enum BridgeRequestStatus {
  CREATED
  PENDING_PRETRANSFER
  PENDING_ON_DESTINATION_CHAIN
  CONFIRMED
  FAILED
  PENDING_WIRON_BURN_TRANSACTION_CREATION
  PENDING_WIRON_BURN_TRANSACTION_CONFIRMATION
  PENDING_IRON_RELEASE_TRANSACTION_CREATION
  PENDING_IRON_RELEASE_TRANSACTION_CONFIRMATION

  @@map("bridge_request_status")
}

enum FailureReason {
  REQUEST_NON_EXISTENT
  REQUEST_INVALID_STATUS
  REQUEST_SOURCE_ADDRESS_NOT_MATCHING
  REQUEST_AMOUNT_NOT_MATCHING
  REQUEST_ASSET_NOT_MATCHING
  WIRON_BURN_TRANSACTION_FAILED

  @@map("failure_reason")
}
