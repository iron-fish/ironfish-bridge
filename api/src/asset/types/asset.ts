/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Chain } from '@prisma/client';

/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
export interface Asset {
  chain: Chain;
  sourceSymbol: string;
  destinationSymbol: string;
  assetId: string;
}

const USDC: Asset = {
  chain: Chain.ETHEREUM,
  sourceSymbol: 'USDC',
  destinationSymbol: 'if.USDC',
  assetId: 'foo',
};

export const SupportedAssets: { [key: string]: Asset } = {
  USDC: USDC,
};
