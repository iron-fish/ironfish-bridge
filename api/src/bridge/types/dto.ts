/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { BridgeRequestStatus, Chain } from '@prisma/client';

type Address = string;

type AddressFk = number;

export type BridgeDataDTO = {
  amount: string;
  source_address: Address;
  destination_address: Address;
  asset: string;
  source_chain: Chain;
  destination_chain: Chain;
  source_transaction: string;
  destination_transaction: string | null;
  source_burn_transaction?: string | null;
  status: BridgeRequestStatus;
};

export type BridgeRetrieveDTO = {
  [keyof: AddressFk]: BridgeDataDTO | null;
};

export type BridgeSendRequestDTO = {
  amount: string;
  asset: string;
  source_address: Address;
  source_chain: Chain;
  source_transaction: string;
  destination_address: Address;
  destination_chain: Chain;
};

export type BridgeSendItemDTO = {
  status: BridgeRequestStatus | null;
  failureReason: string | null;
};

export type BridgeSendResponseDTO = {
  [keyof: AddressFk]: BridgeSendItemDTO;
};

export type BridgeCreateDTO = { [keyof: Address]: AddressFk };

export type HeadHash = { hash: string };

export type OptionalHeadHash = { hash: string | null };

export type UpdateRequestDTO = {
  id: AddressFk;
  destination_transaction: string;
  status: BridgeRequestStatus;
};

export type UpdateResponseDTO = {
  [keyof: AddressFk]: { status: BridgeRequestStatus | null };
};
