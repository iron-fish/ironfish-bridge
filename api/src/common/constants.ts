/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
export const SEPOLIA_EXPLORER_URL = 'https://sepolia.etherscan.io';
export const SEPOLIA_BLOCK_TIME_MS = 15 * 1000;

export const WIRON_CONTRACT_ADDRESS =
  '0x5c6c596cD1866315117d783af9045d9eB8F44802';
