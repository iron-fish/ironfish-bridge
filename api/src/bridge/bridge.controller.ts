/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Post,
  Query,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import {
  BridgeRequest,
  BridgeRequestStatus,
  FailureReason,
} from '@prisma/client';
import assert from 'assert';
import { ApiConfigService } from '../api-config/api-config.service';
import { ApiKeyGuard } from '../auth/guards/api-key.guard';
import { SupportedAssets } from '../common/constants';
import { List } from '../common/interfaces/list';
import { GraphileWorkerPattern } from '../graphile-worker/enums/graphile-worker-pattern';
import { GraphileWorkerService } from '../graphile-worker/graphile-worker.service';
import { MintWIronOptions } from '../wiron/interfaces/mint-wiron-options';
import { BridgeService } from './bridge.service';
import {
  BridgeRetrieveDTO,
  BridgeSendRequestDTO,
  BridgeSendResponseDTO,
  HeadHash,
  OptionalHeadHash,
  UpdateRequestDTO,
  UpdateResponseDTO,
} from './types/dto';
import { NextBridgeRequestsDto } from './types/next-bridge-requests.dto';

@Controller('bridge')
export class BridgeController {
  constructor(
    private readonly config: ApiConfigService,
    private readonly bridgeService: BridgeService,
    private readonly graphileWorkerService: GraphileWorkerService,
  ) {}

  @UseGuards(ApiKeyGuard)
  @Post('retrieve')
  @HttpCode(200)
  async retrieve(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { ids }: { ids: number[] },
  ): Promise<BridgeRetrieveDTO> {
    const requests = await this.bridgeService.findByIds(ids);
    const map: BridgeRetrieveDTO = {};
    for (const id of ids) {
      map[id] = requests.find((r) => r.id === id) ?? null;
    }
    return map;
  }

  @UseGuards(ApiKeyGuard)
  @Post('send')
  async send(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { sends }: { sends: BridgeSendRequestDTO[] },
  ): Promise<BridgeSendResponseDTO> {
    const response = await this.upsertBridgeSendRequestDTOs(
      sends,
      BridgeRequestStatus.PENDING_DESTINATION_MINT_TRANSACTION_CREATION,
    );
    for (const key in response) {
      if (response[key].status === BridgeRequestStatus.FAILED) {
        continue;
      }
      await this.graphileWorkerService.addJob<MintWIronOptions>(
        GraphileWorkerPattern.MINT_WIRON,
        {
          bridgeRequest: Number(key),
        },
      );
    }
    return response;
  }

  @UseGuards(ApiKeyGuard)
  @Post('burn')
  async burn(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { burns }: { burns: BridgeSendRequestDTO[] },
  ): Promise<BridgeSendResponseDTO> {
    const response = await this.upsertBridgeSendRequestDTOs(
      burns,
      BridgeRequestStatus.PENDING_SOURCE_BURN_TRANSACTION_CREATION,
    );
    return response;
  }

  @UseGuards(ApiKeyGuard)
  @Post('release')
  async release(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { releases }: { releases: BridgeSendRequestDTO[] },
  ): Promise<BridgeSendResponseDTO> {
    const response = await this.upsertBridgeSendRequestDTOs(
      releases,
      BridgeRequestStatus.PENDING_DESTINATION_RELEASE_TRANSACTION_CREATION,
    );
    for (const key in response) {
      if (response[key].status === BridgeRequestStatus.FAILED) {
        continue;
      }
      await this.graphileWorkerService.addJob(
        GraphileWorkerPattern.RELEASE_TEST_USDC,
        {
          bridgeRequest: Number(key),
        },
      );
    }
    return response;
  }

  @UseGuards(ApiKeyGuard)
  @Post('head')
  async postHead(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { head }: { head: string },
  ): Promise<HeadHash> {
    const returnedHead = await this.bridgeService.updateHead(head);
    return { hash: returnedHead.hash };
  }

  @UseGuards(ApiKeyGuard)
  @Get('head')
  async getHead(): Promise<OptionalHeadHash> {
    const ethBridgeHead = await this.bridgeService.getHead();
    const head = ethBridgeHead ? ethBridgeHead.hash : null;
    return { hash: head };
  }

  @Get('address')
  getAddress(): { address: string } {
    const address = this.config.get<string>('IRONFISH_BRIDGE_ADDRESS');
    return { address };
  }

  @UseGuards(ApiKeyGuard)
  @Post('update_requests')
  async updateBridgeRequests(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { transactions }: { transactions: UpdateRequestDTO[] },
  ): Promise<UpdateResponseDTO> {
    const requests = await this.bridgeService.findByIds(
      transactions.map((t) => t.id),
    );
    const response: UpdateResponseDTO = {};
    for (const transaction of transactions) {
      const request = requests.find((r) => r.id === transaction.id) ?? null;

      if (!request) {
        response[transaction.id] = { status: null };
        continue;
      }

      await this.bridgeService.updateRequest({
        id: transaction.id,
        status: transaction.status,
        destination_transaction: transaction.destination_transaction,
      });

      response[transaction.id] = { status: transaction.status };
    }
    return response;
  }

  @Get('next_release_requests')
  @UseGuards(ApiKeyGuard)
  async nextReleaseBridgeRequests(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { count }: NextBridgeRequestsDto,
  ): Promise<List<BridgeRequest>> {
    return {
      object: 'list',
      data: await this.bridgeService.nextReleaseBridgeRequests(count),
    };
  }

  @Get('next_burn_requests')
  @UseGuards(ApiKeyGuard)
  async nextBurnBridgeRequests(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    { count }: NextBridgeRequestsDto,
  ): Promise<List<BridgeRequest>> {
    return {
      object: 'list',
      data: await this.bridgeService.nextBurnBridgeRequests(count),
    };
  }

  async upsertBridgeSendRequestDTOs(
    payloads: BridgeSendRequestDTO[],
    status: BridgeRequestStatus,
  ) {
    const response: BridgeSendResponseDTO = {};

    for (const payload of payloads) {
      let request = await this.bridgeService.findBySourceTransaction(
        payload.source_transaction,
      );
      if (!request) {
        let destinationAddress = payload.destination_address;
        if (!destinationAddress.startsWith('0x')) {
          destinationAddress = `0x${destinationAddress}`;
        }

        request = await this.bridgeService.upsertRequest({
          amount: payload.amount,
          asset: payload.asset,
          destination_address: destinationAddress,
          destination_chain: payload.destination_chain,
          destination_transaction: null,
          source_address: payload.source_address,
          source_chain: payload.source_chain,
          source_transaction: payload.source_transaction,
          status,
        });
      }

      let failureReason = request.failure_reason;
      if (!(request.asset in SupportedAssets)) {
        request = await this.bridgeService.updateRequest({
          id: request.id,
          status: BridgeRequestStatus.FAILED,
          failure_reason: FailureReason.REQUEST_ASSET_NOT_MATCHING,
        });
        assert.ok(request);
        await this.bridgeService.createFailedRequest(
          request,
          FailureReason.REQUEST_ASSET_NOT_MATCHING,
          `Request asset is not supported: ${
            request.asset
          }. Supported: ${JSON.stringify(SupportedAssets)} `,
        );
        failureReason = FailureReason.REQUEST_ASSET_NOT_MATCHING;
      }

      response[request.id] = {
        status: request.status,
        failureReason,
      };
    }
    return response;
  }
}
