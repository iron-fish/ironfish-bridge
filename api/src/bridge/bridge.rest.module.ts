/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { GraphileWorkerModule } from '../graphile-worker/graphile-worker.module';
import { BridgeController } from './bridge.controller';
import { BridgeModule } from './bridge.module';

@Module({
  controllers: [BridgeController],
  imports: [ApiConfigModule, BridgeModule, GraphileWorkerModule],
})
export class BridgeRestModule {}
