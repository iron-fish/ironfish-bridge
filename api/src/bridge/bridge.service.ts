/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable } from '@nestjs/common';
import {
  BridgeRequest,
  BridgeRequestStatus,
  FailedBridgeRequest,
  FailureReason,
  IronFishTestnetHead,
  Prisma,
} from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { BasePrismaClient } from '../prisma/types/base-prisma-client';
import { BridgeDataDTO } from './types/dto';

@Injectable()
export class BridgeService {
  constructor(private readonly prisma: PrismaService) {}

  async find(id: number): Promise<BridgeRequest | null> {
    return this.prisma.bridgeRequest.findFirst({
      where: { id },
    });
  }

  async findOrThrow(id: number): Promise<BridgeRequest> {
    return this.prisma.bridgeRequest.findFirstOrThrow({
      where: { id },
    });
  }

  async findByIds(ids: number[]): Promise<BridgeRequest[]> {
    return this.prisma.bridgeRequest.findMany({
      where: {
        id: {
          in: ids,
        },
      },
    });
  }

  async findBySourceTransaction(
    sourceTransaction: string,
  ): Promise<BridgeRequest | null> {
    return this.prisma.bridgeRequest.findUnique({
      where: {
        source_transaction: sourceTransaction,
      },
    });
  }

  async upsertRequests(
    requests: BridgeDataDTO[],
    client?: BasePrismaClient,
  ): Promise<BridgeRequest[]> {
    const results = [];
    for (const request of requests) {
      results.push(await this.upsertRequest(request, client));
    }
    return results;
  }

  async upsertRequest(
    request: BridgeDataDTO,
    client?: BasePrismaClient,
  ): Promise<BridgeRequest> {
    const prisma = client ?? this.prisma;
    return prisma.bridgeRequest.upsert({
      create: {
        ...request,
      },
      update: {
        source_transaction: request.source_transaction,
        destination_transaction: request.destination_transaction,
        status: request.status,
        wiron_burn_transaction: request.wiron_burn_transaction,
      },
      where: {
        source_transaction: request.source_transaction,
      },
    });
  }

  async updateRequest(
    options: {
      id: number;
      status?: BridgeRequestStatus;
      destination_transaction?: string;
      source_transaction?: string;
      wiron_burn_transaction?: string;
    },
    client?: BasePrismaClient,
  ): Promise<BridgeRequest | null> {
    const prisma = client ?? this.prisma;
    return prisma.bridgeRequest.update({
      data: options,
      where: {
        id: options.id,
      },
    });
  }

  async updateHead(hash: string): Promise<IronFishTestnetHead> {
    await this.prisma.ironFishTestnetHead.deleteMany();
    return this.prisma.ironFishTestnetHead.create({
      data: { hash },
    });
  }

  async getHead(): Promise<IronFishTestnetHead | null> {
    return this.prisma.ironFishTestnetHead.findFirst();
  }

  async createFailedRequest(
    request: BridgeRequest | null,
    failureReason: FailureReason,
    error?: string,
    client?: BasePrismaClient,
  ): Promise<FailedBridgeRequest> {
    const prisma = client ?? this.prisma;
    const bridge_request = request
      ? { connect: { id: request.id } }
      : undefined;
    return prisma.failedBridgeRequest.create({
      data: {
        bridge_request,
        error,
        failure_reason: failureReason,
      },
    });
  }

  async nextWIronBridgeRequests(count?: number): Promise<BridgeRequest[]> {
    return this.prisma.bridgeRequest.findMany({
      where: {
        source_chain: 'ETHEREUM',
        destination_chain: 'IRONFISH',
        status: BridgeRequestStatus.PENDING_IRON_RELEASE_TRANSACTION_CREATION,
      },
      orderBy: {
        created_at: Prisma.SortOrder.asc,
      },
      take: count ?? 1,
    });
  }
}
