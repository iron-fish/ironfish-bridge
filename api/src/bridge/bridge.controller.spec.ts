/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
} from '@jest/globals';
import { HttpStatus, INestApplication } from '@nestjs/common';
import { BridgeRequestStatus, FailureReason } from '@prisma/client';
import assert from 'assert';
import request from 'supertest';
import { ApiConfigService } from '../api-config/api-config.service';
import { GraphileWorkerService } from '../graphile-worker/graphile-worker.service';
import { PrismaService } from '../prisma/prisma.service';
import { bridgeRequestDTO } from '../test/mocks';
import { bootstrapTestApp } from '../test/test-app';
import { BridgeService } from './bridge.service';
import { BridgeSendRequestDTO } from './types/dto';

describe('AssetsController', () => {
  let app: INestApplication;
  let prisma: PrismaService;
  let config: ApiConfigService;
  let bridgeService: BridgeService;
  let graphileWorkerService: GraphileWorkerService;
  let API_KEY: string;
  let IRONFISH_BRIDGE_ADDRESS: string;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    prisma = app.get(PrismaService);
    config = app.get(ApiConfigService);
    bridgeService = app.get(BridgeService);
    graphileWorkerService = app.get(GraphileWorkerService);

    API_KEY = config.get<string>('IRONFISH_BRIDGE_API_KEY');
    IRONFISH_BRIDGE_ADDRESS = config.get<string>('IRONFISH_BRIDGE_ADDRESS');

    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  beforeEach(async () => {
    await prisma.bridgeRequest.deleteMany({});
  });

  describe('POST /bridge/create', () => {
    describe('when data is posted to endpoint', () => {
      it('is successful creates record and returns fk', async () => {
        const source_address = '11111111111111111111111111';
        const data = bridgeRequestDTO({ source_address });
        const { body } = await request(app.getHttpServer())
          .post('/bridge/create')
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ requests: [data] })
          .expect(HttpStatus.CREATED);

        const entry = await prisma.bridgeRequest.findFirst({
          where: { source_address: { equals: source_address } },
        });
        assert.ok(entry);

        expect(body).toMatchObject({
          [source_address]: entry.id,
        });
      });
    });
  });

  describe('POST /bridge/retrieve', () => {
    describe('when data is requested by id', () => {
      it('is successfully returns if fk is in db, null if not', async () => {
        const unsavedId = 1234567;
        const requestData = bridgeRequestDTO({});
        const foo = await prisma.bridgeRequest.create({
          data: requestData,
        });
        const { body } = await request(app.getHttpServer())
          .get('/bridge/retrieve')
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ ids: [foo.id, unsavedId] })
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          [foo.id]: {
            ...requestData,
          },
          [unsavedId]: null,
        });
      });
    });
  });

  async function expectedSendFailure(
    send: BridgeSendRequestDTO,
    failureReason: FailureReason,
  ) {
    const response = await request(app.getHttpServer())
      .post('/bridge/send')
      .set('Authorization', `Bearer ${API_KEY}`)
      .send({
        sends: [send],
      })
      .expect(HttpStatus.CREATED);

    expect(response.body).toMatchObject({
      [send.id]: {
        status: BridgeRequestStatus.FAILED,
        failureReason,
      },
    });
  }
  describe('POST /bridge/head', () => {
    describe('updates or creates head for tracking sync progress', () => {
      it('creates then updates head', async () => {
        await request(app.getHttpServer())
          .post('/bridge/head')
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ head: 'fakehash1' })
          .expect(HttpStatus.CREATED);

        const { body } = await request(app.getHttpServer())
          .post('/bridge/head')
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({ head: 'fakehash2' })
          .expect(HttpStatus.CREATED);

        expect(body).toEqual({
          hash: 'fakehash2',
        });

        const count = await prisma.bridgeHead.count();
        expect(count).toBe(1);

        const { body: getBody } = await request(app.getHttpServer())
          .get('/bridge/head')
          .set('Authorization', `Bearer ${API_KEY}`)
          .expect(HttpStatus.OK);

        expect(getBody.hash).toBe('fakehash2');
      });
    });
  });

  describe('POST /bridge/send', () => {
    describe('failure cases', () => {
      it('nonexistent request id fails', async () => {
        const send: BridgeSendRequestDTO = {
          id: 123132132,
          source_transaction: '123123',
          amount: '100',
          asset: 'asdasfsdafdsafdsa',
          source_address: '11111111111111111111111111',
        };
        await expectedSendFailure(send, FailureReason.REQUEST_NON_EXISTENT);
        expect(true).toBe(true);
      });

      it('invalid status fails', async () => {
        const dto = bridgeRequestDTO({
          status: BridgeRequestStatus.PENDING_PRETRANSFER,
        });
        const bridgeRequest = await bridgeService.createRequests([dto]);
        const send: BridgeSendRequestDTO = {
          ...dto,
          source_transaction: dto.source_transaction || '123123',
          id: bridgeRequest[0].id,
        };
        await expectedSendFailure(send, FailureReason.REQUEST_INVALID_STATUS);
        expect(true).toBe(true);
      });

      it('non matching source address fails', async () => {
        const dto = bridgeRequestDTO({});
        const bridgeRequest = await bridgeService.createRequests([dto]);
        const send: BridgeSendRequestDTO = {
          ...dto,
          source_transaction: dto.source_transaction || '123123',
          source_address: 'some different address, not the one in db',
          id: bridgeRequest[0].id,
        };
        await expectedSendFailure(
          send,
          FailureReason.REQUEST_SOURCE_ADDRESS_NOT_MATCHING,
        );
        expect(true).toBe(true);
      });

      it('asset id does not match created entry', async () => {
        const dto = bridgeRequestDTO({});
        const bridgeRequest = await bridgeService.createRequests([dto]);
        const send: BridgeSendRequestDTO = {
          ...dto,
          source_transaction: dto.source_transaction || '123123',
          asset: 'some different address, not the one in db',
          id: bridgeRequest[0].id,
        };
        await expectedSendFailure(
          send,
          FailureReason.REQUEST_ASSET_NOT_MATCHING,
        );
        expect(true).toBe(true);
      });

      it('amount does not match created entry', async () => {
        const dto = bridgeRequestDTO({});
        const bridgeRequest = await bridgeService.createRequests([dto]);
        const send: BridgeSendRequestDTO = {
          ...dto,
          source_transaction: dto.source_transaction || '123123',
          amount: '42069',
          id: bridgeRequest[0].id,
        };
        await expectedSendFailure(
          send,
          FailureReason.REQUEST_AMOUNT_NOT_MATCHING,
        );
        expect(true).toBe(true);
      });
    });

    describe('success case', () => {
      it('updates the request and initiates transfer via smartcontact', async () => {
        const dto = bridgeRequestDTO({});
        const bridgeRequest = await bridgeService.createRequests([
          { ...dto, status: BridgeRequestStatus.CREATED },
        ]);

        const addJobMock = jest
          .spyOn(graphileWorkerService, 'addJob')
          .mockImplementationOnce(jest.fn());

        const send: BridgeSendRequestDTO = {
          ...dto,
          source_transaction: dto.source_transaction || '123123',
          id: bridgeRequest[0].id,
        };

        const response = await request(app.getHttpServer())
          .post('/bridge/send')
          .set('Authorization', `Bearer ${API_KEY}`)
          .send({
            sends: [send],
          })
          .expect(HttpStatus.CREATED);

        expect(addJobMock).toHaveBeenCalledTimes(1);
        const updatedRequest = await bridgeService.findByIds([
          bridgeRequest[0].id,
        ]);

        expect(updatedRequest[0].status).toBe(
          BridgeRequestStatus.PENDING_PRETRANSFER,
        );

        expect(response.body).toMatchObject({
          [bridgeRequest[0].id]: {
            status: BridgeRequestStatus.PENDING_PRETRANSFER,
          },
        });
      });
    });
  });

  describe('GET /bridge/address', () => {
    it('returns the configured public Iron Fish address of the bridge', async () => {
      const { body } = await request(app.getHttpServer())
        .get('/bridge/address')
        .expect(HttpStatus.OK);

      expect(body.address).toEqual(IRONFISH_BRIDGE_ADDRESS);
    });
  });
});
