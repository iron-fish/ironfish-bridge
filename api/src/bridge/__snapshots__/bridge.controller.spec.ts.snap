// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BridgeController GET /bridge/next_wiron_requests with a missing api key returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
